### PR TITLE
Slick 2.1

### PR DESCRIPTION
**Pull request by andrewconner**
Upgrades Slick from version 2.0 to 2.1.
- Supports Scala 2.11, letting us take advantage of some nice improvements in the language.
- Insert-or-update support
- `take` and `drop` can use pre-compiled queries (this will save ~1ms per query)
- Insert queries can be pre-compiled
- Many bug fixes, including more efficient queries.

Biggest eng changes:
- Slick's APIs are closer to Scala collections, so many methods that supported using `()` (such as `toMap`) must be called without the parens. `isNull` is now called `isEmpty`, `isNotNull` is called `isDefined`.
- Nullable columns must be declared as such in the column definition (ie, `column[Option[String]](...)`), to avoid many issues. This is a more-clear, and better semantic.
- `GetResult` and `SetPosition` changed a bit. Typesafe support potentially saved us a day.

As this is a big PR, I'd love to merge soon, but would love some eyes. All tests pass, but we all know how brittle the relationship between SQL → Slick can be.

@stkem @yingjieMiao @leogrim @ahsu1230 @eishay

### CODE REVIEW

**Comment by eishay** [comment]
wohoooo!

File: kifi-backend/modules/abook/app/com/keepit/abook/model/RichSocialConnectionRepo.scala [hunk 1]
```diff
@@ -38,18 +38,18 @@ class RichSocialConnectionRepoImpl @Inject() (
   type RepoImpl = RichSocialConnectionTable
   class RichSocialConnectionTable(tag: Tag) extends RepoTable[RichSocialConnection](db, tag, "rich_social_connection") {
     def userId = column[Id[User]]("user_id", O.NotNull)
-    def userSocialId = column[Id[SocialUserInfo]]("user_social_id", O.Nullable)
+    def userSocialId = column[Option[Id[SocialUserInfo]]]("user_social_id", O.Nullable)
```
Comment (eishay) [inline kifi-backend/modules/abook/app/com/keepit/abook/model/RichSocialConnectionRepo.scala:41]: nice

File: kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/FortyTwoGenericTypeMappers.scala [hunk 2]
```diff
@@ -149,9 +149,34 @@ trait FortyTwoGenericTypeMappers { self: { val db: DataBaseComponent } =>
 
   // SetParameter conversions to be used for interpolated query parameters
 
-  def setParameterFromMapper[T](implicit mapper: BaseColumnType[T]): SetParameter[T] = SetParameter { case (value, parameters) => mapper.setValue(value, parameters) }
-  def setOptionParameterFromMapper[T](implicit mapper: BaseColumnType[T]): SetParameter[Option[T]] = SetParameter { case (option, parameters) => mapper.setOption(option, parameters) }
-  def setSeqParameter[T](implicit pconv: SetParameter[T]): SetParameter[Seq[T]] = SetParameter { case (seq, parameters) => seq.foreach(pconv(_, parameters)) }
+  def setParameterFromMapper[T](implicit mapper: BaseColumnType[T]): SetParameter[T] = {
+    SetParameter {
+      case (value, pp) =>
+        val npos = pp.pos + 1
```
Comment (ahsu1230) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/FortyTwoGenericTypeMappers.scala:155]: what does the +1 do?

File: kifi-backend/modules/abook/app/com/keepit/abook/model/RichSocialConnectionRepo.scala [hunk 3]
```diff
@@ -38,18 +38,18 @@ class RichSocialConnectionRepoImpl @Inject() (
...
+    def friendSocialId = column[Option[Id[SocialUserInfo]]]("friend_social_id", O.Nullable)
+    def friendEmailAddress = column[Option[EmailAddress]]("friend_email_address", O.Nullable)
+    def friendUserId = column[Option[Id[User]]]("friend_user_id", O.Nullable)
+    def friendName = column[Option[String]]("friend_name", O.Nullable)
     def commonKifiFriendsCount = column[Int]("common_kifi_friends_count", O.NotNull)
     def kifiFriendsCount = column[Int]("kifi_friends_count", O.NotNull)
     def invitationsSent = column[Int]("invitations_sent", O.NotNull)
     def invitedBy = column[Int]("invited_by", O.NotNull)
     def blocked = column[Boolean]("blocked", O.NotNull)
-    def * = (id.?, createdAt, updatedAt, state, userId, userSocialId.?, connectionType, friendSocialId.?, friendEmailAddress.?, friendName.?, friendUserId.?, commonKifiFriendsCount, kifiFriendsCount, invitationsSent, invitedBy, blocked) <> ((RichSocialConnection.apply _).tupled, RichSocialConnection.unapply _)
```
Comment (ahsu1230) [inline kifi-backend/modules/abook/app/com/keepit/abook/model/RichSocialConnectionRepo.scala:52]: yayyyy no more `.?`

File: kifi-backend/modules/cortex/app/com/keepit/cortex/dbmodel/URILDATopicRepo.scala [hunk 4]
```diff
@@ -185,15 +185,15 @@ class URILDATopicRepoImpl @Inject() (
...
+
     val q = sql"""select tp.feature from cortex_keep as ck inner join uri_lda_topic as tp on ck.uri_id = tp.uri_id where ck.user_id = ${userId.id} and ck.state = 'active' and tp.version = ${version.version} and ck.source != 'default' and tp.state = 'active' and tp.num_words > ${min_num_words}"""
     q.as[LDATopicFeature].list
   }
 
   def getUserRecentURIFeatures(userId: Id[User], version: ModelVersion[DenseLDA], min_num_words: Int, limit: Int)(implicit session: RSession): Seq[(Id[Keep], Seq[LDATopic], LDATopicFeature)] = {
     import StaticQuery.interpolation
-    import scala.slick.jdbc.GetResult
-    implicit val getFeature = GetResult(r => (r.nextLong(), r.nextLong(), r.nextShort(), r.nextShort(), r.nextShort(), ldaTopicFeatureMapper.nextValue(r)))
+    implicit val getLibraryFeature = getResultFromMapper[LDATopicFeature]
```
Comment (yingjieMiao) [inline kifi-backend/modules/cortex/app/com/keepit/cortex/dbmodel/URILDATopicRepo.scala:196]: why some `implicit` are removed here? would `nextLong()`, `nextShort()` automatically get resolved correctly?

**Comment by eishay** [comment]
good job!

File: kifi-backend/modules/cortex/app/com/keepit/cortex/dbmodel/URILDATopicRepo.scala (hunk 4 above)
Comment (andrewconner) [inline kifi-backend/modules/cortex/app/com/keepit/cortex/dbmodel/URILDATopicRepo.scala:196]: It wasn't necessary. Creating a `GetResult` from a given mapper is a bit more complicated now, and has been done in `getResultFromMapper`.

File: kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/FortyTwoGenericTypeMappers.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/FortyTwoGenericTypeMappers.scala:155]: Honestly, I see what it's doing (imperatively pulling the next field out), but wouldn't have been able to come up with this. Stefan Zeiger, maintainer of Slick (by our Typesafe support), wrote it for us.

Comment (yingjieMiao) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/FortyTwoGenericTypeMappers.scala:155]: Haha, Zeiger wrote this:

https://github.com/slick/slick/blob/2.1.0/src/main/scala/scala/slick/jdbc/PositionedResult.scala

**Comment by andrewconner** [comment]
Will merge when there's more people in the office.

**Comment by 2is10** [comment]
:+1:

**Pull request merged**